### ISSUE TITLE
Superfluous or too specialized code needs to be removed

### ISSUE THREAD

**Issue by clinuxrulz** [DanielGronau/highj#17]

**Comment by DanielGronau** [comment]
Starting with collections:
- Dequeue seems to be useful (if it is implemented correctly)
- Concerning HList I'm not sure
- MultiSet is not really necessary 
- the functionality of Stream could be replaced by using Lists, but we would lose the StreamMonad (which is "zippy", not "crossy"), and the type wouldn't communicate any longer that it is an infinite structure. If we keep Stream, the common interface with List should be extracted 
- Tree isn't terrible useful right now, could be dropped

**Comment by clinuxrulz** [comment]
I'm not 100% sure with collections, I might leave that up to you. Its true that Streams and Lists overlap a lot. The distinction in FunctionJava is that Lists are Strict and Streams are Lazy. In PureScript they distinguish them by package name ```Data.List``` and ```Data.List.Lazy```, since Java does not support ```import ... as``` construct we might need different names to do the same thing.

One thing we can drop is the ```operational``` package containing ```ProgramT``` and ```ProgramViewT```. As Scala has a better stack-less implementation for their ```Free``` / ```FreeT```. And I probably should have used ```Free``` / ```FreeT``` for class names as it is the Free-monad anyway. Free-monads come in handy if a user wants to build their own monad when the implementation seems to complicated (Free-monads can simplify the implementation, and also allows multiple interpretations of the same computation).

**Comment by jbgi** [comment]
be the way, about strict vslLazy list. One approach could be to support both with the same type, using a lazy constructor. cf. https://github.com/derive4j/derive4j#first-class-laziness and the implementation generated in https://gist.github.com/jbgi/43c1bd0ab67e3f4b9634#file-lists-java-L99

Support for highj is planned for the next derive4j release, so maybe it could also help, in particular to provide true lazyness (with memoization)...

**Comment by DanielGronau** [comment]
The current List implementation already supports strict and lazy behavior, although it is not very consistent regarding which operation behaves in which way.

**Comment by jbgi** [comment]
Yes, though the current ```newLazyList``` constructor is still strict in 2 ways: 
1) caller must eagerly know that he/she want to construct a non empty list (which could require strict evaluation). and even then only the tail is call-by-name. 
2) even then the tail it is not lazy as it is not memoized.

Another "problem" with the current implementation is that the representation is a product type (of the two abstract methods): an implementation of the contract could return a ```Just``` for ```maybeHead()``` and a ```Nothing``` for ```maybeTail```.

The code generated by derive4j address all those "problems" for free.

**Comment by clinuxrulz** [comment]
Java is a very verbose language. Automatic code-gen is very handy.

@DanielGronau I'd recommend making derive4j a dependency of highj, if your OK with that.

**Comment by jbgi** [comment]
And derive4j is compile only (not a runtime dependencies)

**Comment by DanielGronau** [comment]
@clinuxrulz Yes, of course.

**Comment by clinuxrulz** [comment]
Sadly ```Stream``` from FunctionalJava is not fully lazy either, and it is too late for them to fix it.

**Comment by clinuxrulz** [comment]
@DanielGronau Actually there is a tiny issue with using derive4j for defining ```List```. Its that the users of the library would have to go ```Lists.Cons(...)``` and ```Lists.Nil```, Lists with an "s". This is not really a big problem as long as it is documented. Just thought I'd make you aware of it.

Incase you decide not to define ```List``` using derive4j, then constructing a true lazy list would be more like ```lazyList(Lazy<List<A>> a)``` where ```Lazy``` is like a ```Supplier``` than caches/remembers its value after its first call (a bit different to how Lazy is currently defined in highj). That is for a list to be fully lazy, it must not be known ahead of time whether the list is empty or non-empty.

The same thing goes for ```Either```. A true lazy ```Either``` can not know whether it is left or right ahead of time. So a lazy ```Either``` is more like ```lazyEither(Lazy<Either<A,B>> a)```.

Of course derive4j can do all that for us for free. Its just a matter of if your ok with ```Eithers.Right(...)``` as apposed to ```Either.Right(...)```. Check the derive4j documentation out if you have time.

derive4j also generates ```Lens``` and mixes well with optics i think.

**Comment by jbgi** [comment]
@clinuxrulz I would actually recommend to not expose derive4j generated methods as public, so that you are more free to evolve highj api whithout breaking compatibility: you can configure the visibility and the name of the generated class, eg.:
```java
@Data(@Derive(inClass="ListImpl", withVisibility=Package))
```
And then you delegate to ```ListImpl``` from ``List``` (or other same package class).

**Comment by clinuxrulz** [comment]
@jbgi Thanks, that sounds like a good plan.

**Comment by clinuxrulz** [comment]
@jbgi I need just a little bit of help. How do I use ```flavour``` and ```@Derive``` together?

Example: ```@Data(@Derive(inClass="ListImpl", withVisibility=Visibility.Package), flavour = Flavour.FJ)```

I can not get the above to compile.

**Comment by jbgi** [comment]
in presence of other attributes, the default attribute must be explicitly named:
```@Data(value = @Derive(inClass="ListImpl", withVisibility=Visibility.Package), flavour = Flavour.FJ)```

**Comment by clinuxrulz** [comment]
Ah! Cheers.
